addons/adyen_marketpay/models/adyen_account.py

```diff
@@ -0,0 +1,640 @@
+    ], required=True, default='awaiting_data')
+    kyc_error_message = fields.Char('KYC Error Message', readonly=True)
+
+    _sql_constraints = [
+        ('adyen_uuid_uniq', 'UNIQUE(adyen_uuid)', 'Adyen UUID should be unique'),
+    ]
+
+    @api.depends('transaction_ids')
+    def _compute_transactions_count(self):
+        for adyen_account_id in self:
+            adyen_account_id.transactions_count = len(adyen_account_id.transaction_ids)
+
+    @api.depends('first_name', 'last_name', 'legal_business_name')
+    def _compute_full_name(self):
+        for adyen_account_id in self:
+            if adyen_account_id.is_business:
+                adyen_account_id.full_name = adyen_account_id.legal_business_name
+            else:
+                adyen_account_id.full_name = "%s %s" % (adyen_account_id.first_name, adyen_account_id.last_name)
+
+    @api.model
+    def create(self, values):
+        adyen_account_id = super(AdyenAccount, self).create(values)
+        self.env.company.adyen_account_id = adyen_account_id.id
+
+        # Create account on odoo.com, proxy and Adyen
+        response = adyen_account_id._adyen_rpc('create_account_holder', {
+            **adyen_account_id._format_data(),
+            'processingTier': 1,
+        })
+
+        # Save adyen_uuid and proxy_token, that have been generated by odoo.com and the proxy
+        adyen_account_id.with_context(update_from_adyen=True).write({
+            'adyen_uuid': response['adyen_uuid'],
+            'proxy_token': response['adyen_proxy_token'],
+        })
+
+        # A default payout is created for all adyen accounts
+        adyen_account_id.env['adyen.payout'].with_context(update_from_adyen=True).create({
+            'code': response['adyen_response']['accountCode'],
+            'adyen_account_id': adyen_account_id.id,
+        })
+        return adyen_account_id
+
+    def write(self, vals):
+        res = super(AdyenAccount, self).write(vals)
+        if not self.env.context.get('update_from_adyen'):
+            self._adyen_rpc('update_account_holder', self._format_data())
+        return res
+
+    def unlink(self):
+        for adyen_account_id in self:
+            adyen_account_id._adyen_rpc('close_account_holder', {
+                'accountHolderCode': adyen_account_id.account_holder_code,
+            })
+        return super(AdyenAccount, self).unlink()
+
+    @api.model
+    def action_create_redirect(self):
+        '''
+        Accessing the FormView to create an Adyen account needs to be done through this action.
+        The action will redirect the user to accounts.odoo.com to link an Odoo user_id to the Adyen
+        account. After logging in on odoo.com the user will be redirected to his DB with a token in
+        the URL. This token is then needed to create the Adyen account.
+        '''
+        if self.env.company.adyen_account_id:
+            raise UserError(_('You can only create a single Adyen account per company.'))
```

Comment:
Maybe it should also be in the create (and so forget the comment I made there)

Reply:
`create` should always be called after this action as we need the creation token. Otherwise, the `create` will fail. Do we really need this check twice?

Reply:
no its ok